[DC-208] Fix run-catalog-workflow test

samanehsan opened the pull request:
Changes:
- Check for text on the workspace page instead of the url
- Fix selecting a workspace
- Throw a more descriptive error when creating and deleting a workspace (otherwise it just says `Evaluation failed: Response` and it's difficult to tell what caused it)

samanehsan committed:
Catch error deleting workspace

Catch and log this error so that it
does not obscure the original error
from the test.

samanehsan committed:
CheckBucketAccess after importing to workspace

samanehsan commented on integration-tests/tests/run-catalog-workflow.js:
```diff
@@ -1,28 +1,26 @@
   await click(page, `//*[@role="combobox"][contains(normalize-space(.), "${workspaceName}")]`)
   await click(page, clickable({ textContains: 'Import' }))
   await waitForNoSpinners(page)
+  await checkBucketAccess(page, billingProject, workspaceName)
```
This check actually might not be necessary since the workspace is created at the start of the test, but we could leave it in case it takes a while for the access to propagate

cahrens replied:
It's definitely not useful here because the "Import" isn't actually getting clicked on (that's what I see when I run the test on dev locally with Chrome visible). The page URL check won't actually fail the test.

See [my PR](https://github.com/DataBiosphere/terra-ui/pull/2811/files#r803768256) where I'm commenting out these lines because the clickable xpath being used for "Import" is changing in my PR to require the button to be enabled. It then causes this test to fail because Import is not actually enabled (and thus the click is a no-op).

cahrens replied:
I'm also curious what failure signature you are trying to solve with waiting for bucket access.

samanehsan replied:
Thanks, I'll take a look! Since we're creating the workspace at the start of the test, it seems like there is enough time for the bucket access to propagate but I'm not sure how reliable that is.

cahrens replied:
@samanehsan my PR has merged, so you should rebase your branch to pick up the changes and test your new code with the changes that have been made to clickable.

cahrens commented on integration-tests/utils/integration-helpers.js:
```diff
@@ -37,11 +37,15 @@ const makeWorkspace = withSignedInPage(async ({ page, billingProject }) => {
-  rawConsole.info(`deleted workspace: ${workspaceName}`)
+    rawConsole.info(`deleted workspace: ${workspaceName}`)
+  } catch (e) {
+    console.error(`Failed to delete workspace: ${workspaceName}. Error: ${e.message}`)
```
How would someone know to look for this? Should we fail the test?

Note also we typically use rawConsole, but perhaps the usage of console here was intentional to make this stand out more (you will get a bit of a stack trace)?

samanehsan replied:
I added this in for debugging since I thought the delete was failing and obscuring an actual test error. It turns out the `Evaluation failed` error was in creating a workspace (seen 1/100 times) so I updated that to throw a more descriptive error message and changed the delete workspace function to do the same thing.

samanehsan committed:
Fix run-catalog-workflow

Fix selecting the workspace and
check for the workspace name on
the page instead of in the url

samanehsan committed:
Improve error handling

cahrens commented on integration-tests/tests/run-catalog-workflow.js:
```diff
@@ -10,23 +10,15 @@ const testCatalogFlowFn = _.flow(
 )(async ({ page, testUrl, token, workspaceName }) => {
   await enableDataCatalog(page, testUrl, token)
```
The first 6 lines of this test are identical to lines 7-12 from `link-to-new-workspace`. I would recommend that you share the code between the two tests.

cahrens commented on integration-tests/tests/run-catalog-workflow.js:
```diff
@@ -10,23 +10,15 @@ const testCatalogFlowFn = _.flow(
-  // await page.url().includes(workspaceName)
+  await select(page, 'Select a workspace', `${workspaceName}`)
+  await noSpinnersAfter(page, { action: () => click(page, clickable({ text: 'Import' })) })
+  await findText(page, workspaceName)
```
I would suggest looking for something a bit more specific than just the workspace name, because I would imagine that the workspace name can appear on a lot of pages. One possibility is the combination of the billing project and workspace name, which is what we show in the top context bar (though even that doesn't feel highly specific to the page to make sure loaded). `await findText(page, `${billingProject}/${workspaceName}`)`, which requires you to pass the `billingProject` into the test.

Is there any text or element on the page you expect to load that is definitely unique to that page?

samanehsan replied:
There isn't much else on the page:
<img width="1208" alt="Screen Shot 2022-02-10 at 4 09 49 PM" src="https://user-images.githubusercontent.com/6414394/153496884-16690dee-91b7-49b9-8b27-09aaf51c56ac.png">

samanehsan replied:
How about `await findText(page, ${billingProject}/${workspaceName})` and `await findText(page, 'Workspace Data')` since we expect to be in the data tab?

cahrens replied:
The combination of `await findText(page, ${billingProject}/${workspaceName})` and looking for "Select a data type" would be good.

cahrens commented:
I'm running this with the flaky runner and getting a lot of this error signature (7/40):

````
TimeoutError: waiting for XPath `(//input[@type="checkbox"] | //*[@role="checkbox"])//*[normalize-space(.)="Granted" or @title="Granted" or @aria-label="Granted" or @aria-labelledby=//*[normalize-space(.)="Granted"]/@id]` failed: timeout 30000ms exceeded

samanehsan commented:
@cahrens what kind of computer are you using? I'm on a 2017 macbook pro with 16GB of memory so I've been running the tests with `CONCURRENCY=5`. If I use the default setting to run 10 concurrent tests a bunch of them timeout.

cahrens commented:
I have a 2021 MacBook Pro 32GB of memory. I'm using the default concurrency option of 10, which is what I've been using for all of my flaky testing. I've been able to get clean runs (out of 100) on other tests, so I don't necessarily think this is related to the computer's capabilities.

samanehsan commented:
> I have a 2021 MacBook Pro 32GB of memory. I'm using the default concurrency option of 10, which is what I've been using for all of my flaky testing. I've been able to get clean runs (out of 100) on other tests, so I don't necessarily think this is related to the computer's capabilities.

Okay, that's good to know 🤔

kyuksel commented on integration-tests/utils/integration-helpers.js:
```diff
@@ -26,22 +26,29 @@ const getTestWorkspaceName = () => `test-workspace-${Math.floor(Math.random() *
+      return window.Ajax().Workspaces.create({ namespace: billingProject, name, attributes: {} })
+    }, workspaceName, billingProject)
 
+    rawConsole.info(`created workspace: ${workspaceName}`)
```
Nit:

```suggestion
    rawConsole.info(`Created workspace: ${workspaceName}`)
```

since we typically initialize the first letter of a sentence.

Same for `deleteWorkspace` below.

kyuksel commented on integration-tests/utils/integration-helpers.js:
```diff
@@ -26,22 +26,29 @@ const getTestWorkspaceName = () => `test-workspace-${Math.floor(Math.random() *
 
+    rawConsole.info(`created workspace: ${workspaceName}`)
+  } catch (e) {
+    throw Error(`Failed to create workspace: ${workspaceName}`)
```
Would it be helpful to include `billingProject` in the error messages above and below?

kyuksel reviewed (commented):
Looks good overall. Pending [Christina's comment](https://github.com/DataBiosphere/terra-ui/pull/2814#discussion_r804100542).

kyuksel commented:
> what kind of computer are you using?

FWIW, the specs below are for my Mac, which is more similar to @samanehsan's. I have also been able to run `100` tests with concurrency set to `10`.

![image](https://user-images.githubusercontent.com/5438223/153612049-1c4386c4-471b-4605-b4fb-c48f0a18ace7.png)

samanehsan committed:
Address code review comments

cahrens commented on integration-tests/utils/integration-helpers.js:
```diff
@@ -192,11 +192,19 @@ const enableDataCatalog = async (page, testUrl, token) => {
   await dismissNotifications(page)
 }
 
+const goToLinkWorkspacePage = async (page) => {
```
I was imagining you would keep this within one of the tests that uses it (I don't think it's general enough to go in integration-helpers). And you can put `await enableDataCatalog(page, testUrl, token)` in it also.

samanehsan replied:
Oh I see! I find keeping `enableDataCatalog` separate makes the test read better but I'm fine either way

cahrens replied:
Maybe call it something like "linkDataToWorkspace"? Then it would make sense that it includes enabling the data catalog (which it is interacting with in the test anyway).

cahrens commented on integration-tests/tests/run-catalog-workflow.js:
```diff
@@ -1,24 +1,20 @@
   withWorkspace,
   withUserToken
-)(async ({ page, testUrl, token, workspaceName }) => {
+)(async ({ page, testUrl, token, workspaceName, billingProject }) => {
```
I think we try to keep the variables in alphabetical order.

samanehsan committed:
More clean-up

samanehsan commented:
@kyuksel that's good to know! I have a 3.1 GHz Quad-Core Intel Core i7 processor

samanehsan commented:
@kyuksel @cahrens let me know if the changes look good!

cahrens commented on integration-tests/tests/link-to-new-workspace.js:
```diff
@@ -21,6 +16,7 @@ const testLinkToNewWorkspaceFn = withUserToken(async ({ testUrl, page, token })
+    await findText(page, 'Select a data type')
   } finally {
     try {
       await page.evaluate((name, billingProject) => {
```
Can't you just call `deleteWorkspace` here?

cahrens replied:
If for some reason you cannot, you probably want to throw an error if the workspace can't be deleted, like you did in `deleteWorkspace`.

samanehsan replied:
If the test fails to create a workspace, this delete throws an error. If it's not handled that error obscures the original test error and makes it difficult to debug. Although now that I think of it I could check if the workspace exists before calling `deleteWorkspace` to avoid that situation.

cahrens replied:
If a workspace isn't created, won't the test fail earlier? checkBucketAccess will error I believe, and you shouldn't be able to navigate to the workspace page.

samanehsan replied:
The finally block runs even if there's an error thrown in the try block, so the original error gets lost

samanehsan replied:
This comment from Pete has a useful code snippet to help observe this behavior: https://github.com/DataBiosphere/terra-ui/pull/2807#pullrequestreview-875241605

cahrens replied:
I still think you can call `deleteWorkspace` and just catch any errors that it throws. That way you would not duplicate the deleteWorkspace code.

petesantos replied:
@cahrens Are you referring to [this deleteWorkspace](https://github.com/DataBiosphere/terra-ui/blob/bf9423b28bcf8d0098c076e61e8dd28f5b7f62f3/integration-tests/utils/integration-helpers.js#L42) function?

If so, I tried out this function. It doesn't look like a super-clean abstraction and ends up running some additional code around sign in that I don't think we want to do. We could add a ticket to investigate creating a cleaner utility for this.

cahrens replied:
OK, I was hoping it would be straightforward to use, but if not feel free to punt it.

samanehsan replied:
@petesantos and I paired on improving the error handling here! Now if the test fails it will throw the test failure (and log the cleanup error if that fails too). If the test succeeds and the cleanup fails, it throws the cleanup error.

cahrens commented:
> I'm running this with the flaky runner and getting a lot of this error signature (7/40):
> 
> ```
> TimeoutError: waiting for XPath `(//input[@type="checkbox"] | //*[@role="checkbox"])//*[normalize-space(.)="Granted" or @title="Granted" or @aria-label="Granted" or @aria-labelledby=//*[normalize-space(.)="Granted"]/@id]` failed: timeout 30000ms exceeded
> ```

@samanehsan will you continue to investigate this failure signature ^ for a future PR? I agree with iteratively improving instead of trying to figure out every possible issue at once, but just want to make sure this does get followed up on.

samanehsan commented:
@cahrens yes, I'll make a follow-up ticket for that!

petesantos commented on integration-tests/tests/run-catalog-workflow.js:
```diff
@@ -36,4 +31,7 @@ const testCatalog = {
 
-module.exports = { testCatalog }
+module.exports = {
+  linkDataToWorkspace,
```
I would rather not have the test export a utility function. I think this confuses the purpose of the file - it becomes both test and utility. We could create a `catalog-utils.js` file for this type of code

samanehsan replied:
Ah okay! I think that's a good compromise here

cahrens replied:
You forgot to remove the export of `linkDataToWorkspace` here.

samanehsan replied:
Ah good catch! Just removed it

petesantos commented on integration-tests/utils/integration-helpers.js:
```diff
@@ -26,22 +26,29 @@ const getTestWorkspaceName = () => `test-workspace-${Math.floor(Math.random() *
 
+    rawConsole.info(`Created workspace: ${workspaceName}`)
+  } catch (e) {
+    throw Error(`Failed to create workspace: ${workspaceName} with billing project ${billingProject}`)
```
I think these error messages will be much clearer when they fail - thanks!

samanehsan replied:
I saw this test fail 2/100 times because it failed to create a workspace. Is this something that would get covered by the retry fix you wanted to do? If not I can make a separate ticket for this.

petesantos replied:
Yes, but I think retry is out of scope for this ticket. I think the improvements here will reduce a lot of failure noise / flakiness

kyuksel replied:
I wonder if it makes sense to investigate why workspace creation fails before adding a retry, to make sure retry wouldn't conceal a back-end issue.

petesantos reviewed (commented):
This is looking good.
I ran run-catalog-workflow 100 times with 1 error, a ToS error, that is more systemic and out of scope for the test.
I ran link-to-new-workspace with the same result

petesantos commented:
> > I'm running this with the flaky runner and getting a lot of this error signature (7/40):
> > ```
> > TimeoutError: waiting for XPath `(//input[@type="checkbox"] | //*[@role="checkbox"])//*[normalize-space(.)="Granted" or @title="Granted" or @aria-label="Granted" or @aria-labelledby=//*[normalize-space(.)="Granted"]/@id]` failed: timeout 30000ms exceeded
> > ```
> 
> @samanehsan will you continue to investigate this failure signature ^ for a future PR? I agree with iteratively improving instead of trying to figure out every possible issue at once, but just want to make sure this does get followed up on.

FWIW I got that as well, but it looks like a ToS error from the screenshot:
![98](https://user-images.githubusercontent.com/50371603/153651540-f5ce47eb-bcc5-4cc9-9401-03354c928c7e.jpg)

samanehsan commented:
I created a ticket for the timeout error here: https://broadworkbench.atlassian.net/browse/DC-228

petesantos committed:
Saman/Pete pair

samanehsan committed:
Move helper fns to catalog-utils

samanehsan committed:
Remove linkDataToWorkspace export
